core/station/impl/src/services/external_canister.rs

```diff
@@ -517,196 +519,246 @@ impl ExternalCanisterService {
+                    calls,
+                ) => {
+                    // first aggregates the calls by the execution method
+                    let calls_by_execution_method = calls.iter().fold(
+                        HashMap::new(),
+                        |mut acc: HashMap<
+                            String,
+                            Vec<ExternalCanisterCallRequestPolicyRuleInput>,
+                        >,
+                         call| {
+                            acc.entry(call.execution_method.clone())
+                                .or_default()
+                                .push(call.clone());
+                            acc
+                        },
+                    );
 
-                        continue;
+                    for (method, calls) in calls_by_execution_method {
+                        self.maybe_mutate_canister_calls_request_policies(
+                            external_canister,
+                            &self
+                                .request_policy_repository
+                                .find_external_canister_call_policies_by_execution_method(
+                                    &external_canister.canister_id,
+                                    &method,
+                                ),
+                            &calls,
+                        )?;
                     }
-
-                    self.request_policy_service.edit_request_policy(
-                        EditRequestPolicyOperationInput {
-                            policy_id,
-                            rule: Some(policy.rule),
-                            specifier: None,
-                        },
-                    )?;
                 }
-                None => {
-                    self.request_policy_service.add_request_policy(
-                        AddRequestPolicyOperationInput {
-                            rule: policy.rule,
-                            specifier: RequestSpecifier::ChangeExternalCanister(
-                                ExternalCanisterId::Canister(external_canister.canister_id),
-                            ),
-                        },
-                    )?;
+                ExternalCanisterChangeCallRequestPoliciesInput::RemoveByPolicyIds(
+                    policy_ids_to_remove,
+                ) => {
+                    // removes the specified policies if they are of the call type and exist for the external canister
+                    for policy_id in policy_ids_to_remove {
+                        if !current_policies.calls.contains(policy_id) {
+                            return Err(ExternalCanisterError::ValidationError {
+                                info: format!(
+                                    "The policy with id {} does not exist for the external canister.",
+                                    Uuid::from_bytes(*policy_id).hyphenated()
+                                ),
+                            })?;
+                        }
```

Comment:
Will the subsequent removal not fail anyway (in other words, do we need this explicit check here)?

Reply:
We do, because this validates that the user is not providing a request policy id that is not related to this external canister, otherwise they would be able to make this request and update any other policy in the system

Reply:
Oh, I see. Shouldn't this validation be performed by the authorization middleware?

Reply:
We agreed that it makes sense to have the check at both places: here and in the authorization middleware.